content/concepts/address-discovery-and-advertisement.md

```diff
@@ -0,0 +1,256 @@
+  some of them could be loopback addresses etc.
+
+- As mentioned above, before we share such observed addresses with
+  the world, we need a fair degree of confidence that they are indeed dialable and reachable from outside.
+  We need to do some analytics here and this is where the `ObservedAddrManager` component comes in.
+
+#### The Observed Address Manager
+
+- The purpose of the Observed Address Manager is to help us determine which observed address should be shared
+with the world.
+
+- Every observed address we get using the Identify protocol is recorded in the Observed Address Manager.
+
+- When the Host wants to build it’s address set, one of the steps it takes is to ask the Observed Address Manager
+  for the addresses it thinks should be shared.
+
+- Here are the rules used by the Manager to determine which addresses will be shared:
+    - The observed address should NOT be a loopback address.
+
+    - The local address we see on the connection that led to this observation should be one of the network interface
+      addresses we are listening on. This ensures we don’t share observations that were created by the use of
+      ephemeral ports for outbound connections.
+      We want to ensure the NAT port mapping for the observed address is the same as that for one of our interface
+      listen addresses.
+
+    - The same address should have been observed by 4 different peers (different here means different IP addresses)
+      within the past 40 minutes.
+
+    - The address has been seen at least once in the last 10 minutes(offers some protection against changes in our
+      own network conditions).
+
+    - On top of all the above, the observed address manager selects ONLY the “top” two observed addresses for
+      each group of (IP address + Transport protocol). For ranking, we prefer observations created because of
+      inbound connections over outbound ones(obviously), using the number of peers who reported that same
+      observed address to resolve ties.
+
+#### Relay Addresses: Using AutoRelay, AutoNAT and Relay Servers
+
+- We use [circuit relays][spec_relay] to enable other peers to connect to us when we discover we are completely
+  NATT’d and are not dialable at all. The basic idea is that we connect to a publicly reachable “relay” server and
+  then create a “relay address” which we can advertise.
+  A relay address is usually a combination of the Relay server’s address combined with it’s peer id.
+  Any peer that then wants to talk to us can do so via the relay server.
+
+- There are three parts to this:
+    - Discovering we are completely NATT’d and thus NOT dialable at all using `AutoNAT`.
+
+    - Discovering Relay Servers we can connect to and building relay addresses by using `AutoRelay`.
+
+    - Creating long lived connections with the Relay server and advertising relay addresses in `AutoRelay`.
+
+    #### AutoNAT: Discovering we are NOT dialable/NATT’d
```

Comment:
Formatting looks like it's messed up on these headers due to the indenting. Should probably also be a level 5 header.

Reply:
Have revisited and fixed all headers.